#32: Updated user guide with additional Kerberos guide (#33)
exasol/hive-virtual-schema: doc/user_guide/hive_user_guide.md

```

#### Using Correct `krb5.conf` Configuration File

The regular Kerberos configuration `/etc/krb5.conf` file may contain `includedir` directives that include additional properties. To accommodate all properties in a single file, please make a copy of `/etc/krb5.conf` file and update it accordingly.

Example of `/etc/krb5.conf` contents:

```
includedir /etc/krb5.conf.d/
includedir /var/lib/sss/pubconf/krb5.include.d/

[libdefaults]
dns_lookup_kdc = false
dns_lookup_realm = false
...
```

Copied and changed `krb5.conf` file after adding properties from `includedir` directories:

```
# includedir /etc/krb5.conf.d/
# includedir /var/lib/sss/pubconf/krb5.include.d/

[libdefaults]
dns_lookup_kdc = false
dns_lookup_realm = false
udp_preference_limit = 1

[capaths]
...
[plugins]
...
```

The `includedir` folders contain a file with a setting `[libdefaults]` setting `udp_preference_limit = 1`, two new categories `[capaths]` and `[plugins]` that were not in the original `krb5.conf` file. We add such settings into the copied `krb5.conf` file and use it to create the connection object.

> This is required since the Virtual Schema run in a sandboxed container that does not have an access to the host operating system's files.

Reviewer: ```suggestion
This is required since the Virtual Schema runs in a sandboxed container that does not have an access to the host operating system's files.
```

> Next, we add the JDBC connection URL to the `TO` part of the connection string.

Reviewer 2: ```suggestion
Next, we add the JDBC connection URL to the `TO` part of the connection object.
```

```sql
CREATE OR REPLACE CONNECTION krb_conn
TO 'jdbc:hive2://<Hive host>:<port>;AuthMech=1;KrbAuthType=1;KrbRealm=EXAMPLE.COM;KrbHostFQDN=_HOST;KrbServiceName=hive;transportMode=http;httpPath=cliservice'
USER 'krbuser@EXAMPLE.COM'
IDENTIFIED BY 'ExaAuthType=Kerberos;enp6Cg==;YWFhCg=='
```

#### Kerberos Authentication Type

> As you can see we are using `KrbAuthType=1`. This allows the Hive JDBC driver to check the `java.security.auth.login.config` system property for a JAAS configuration. If a JAAS configuration is specified, the driver uses that information to create a `LoginContext` and then uses the `Subject` associated with it. Exasol Virtual Schema creates `JAAS` configuration at the runtime using the information from the `IDENTIFIED BY` part of the connection object.

Reviewer: ```suggestion
As you can see we are using `KrbAuthType=1`. This allows the Hive JDBC driver to check the `java.security.auth.login.config` system property for a JAAS configuration. If a JAAS configuration is specified, the driver uses that information to create a `LoginContext` and then uses the `Subject` associated with it. Exasol Virtual Schema creates a `JAAS` configuration at runtime using the information from the `IDENTIFIED BY` part of the connection object.
```

   SCHEMA_NAME     = '<schema name>';
```

### Enabling Logging

You can add `LogLevel` and `LogPath` parameters to the JDBC URL (the `TO` part of connection object) to enable Hive JDBC driver and connection logs.

For example:

```
TO 'jdbc:hive2://<Hive host>:<port>;AuthMech=1;KrbAuthType=1;KrbRealm=EXAMPLE.COM;KrbHostFQDN=_HOST;KrbServiceName=hive;transportMode=http;httpPath=cliservice;LogLevel=6;LogPath=/tmp/'
```

> This will create log files in the database `/tmp/` folder. Please check the `exasol_container_sockets` folder during the Virtual Schema creation or query run.

Reviewer: Could you add a reference to a place where the meaning and available values for LogLevel are explained?

   SCHEMA_NAME     = '<schema name>';
```

### Enabling Logging

You can add `LogLevel` and `LogPath` parameters to the JDBC URL (the `TO` part of connection object) to enable Hive JDBC driver and connection logs.

For example:

```
TO 'jdbc:hive2://<Hive host>:<port>;AuthMech=1;KrbAuthType=1;KrbRealm=EXAMPLE.COM;KrbHostFQDN=_HOST;KrbServiceName=hive;transportMode=http;httpPath=cliservice;LogLevel=6;LogPath=/tmp/'
```

> This will create log files in the database `/tmp/` folder. Please check the `exasol_container_sockets` folder for `HiveJDBC_driver.log` and `HiveJDBC_connection_*.log` files during the Virtual Schema creation or query run. These files will contain exception logs if there is any.

Reviewer 2: ```suggestion
This will create log files in a sandbox `/tmp/` folder. Please check the `exasol_container_sockets` folder for `HiveJDBC_driver.log` and `HiveJDBC_connection_*.log` files during the Virtual Schema creation or query run. These files will contain exception logs if there is any.
```

TO 'jdbc:hive2://<Hive host>:<port>;AuthMech=1;KrbAuthType=1;KrbRealm=EXAMPLE.COM;KrbHostFQDN=_HOST;KrbServiceName=hive;transportMode=http;httpPath=cliservice;LogLevel=6;LogPath=/tmp/'
```

This will create log files in a sandbox `/tmp/` folder. Please check the `exasol_container_sockets` folder for `HiveJDBC_driver.log` and `HiveJDBC_connection_*.log` files during the Virtual Schema creation or query run. These files will contain exception logs if there is any.

> You can set the `LogLevel` to number from `0` (Disables all logging) to `6` (Logs all driver activity). For more information please checkout the [Cloudera Hive JDBC Manual](https://docs.cloudera.com/documentation/other/connectors/hive-jdbc/latest/Cloudera-JDBC-Driver-for-Apache-Hive-Install-Guide.pdf) section on "Configuring Logging".

Reviewer: 👍 👍 👍 👍 👍 👍 👍
